lisa/tests/base.py

```diff
@@ -131,28 +166,94 @@ def from_bool(cls, cond, *args, **kwargs):
         result = Result.PASSED if cond else Result.FAILED
         return cls(result, *args, **kwargs)
 
-    def __bool__(self):
-        return self.result is Result.PASSED
+class AggregatedResultBundle(ResultBundleBase):
+    """
+    Aggregates many :class:`ResultBundle` into one.
 
-    def __str__(self):
-        return self.result.name + ': ' + ', '.join(
-                '{}={}'.format(key, val)
-                for key, val in self.metrics.items())
+    :param result_bundles: List of :class:`ResultBundle` to aggregate.
+    :type result_bundles: list(ResultBundle)
 
-    def add_metric(self, name, data, units=None):
-        """
-        Lets you append several test :class:`TestMetric` to the bundle.
+    :param name_metric: Metric to use as the "name" of each result bundle.
+        The value of that metric will be used as top-level key in the
+        aggregated metrics. If not provided, the index in the
+        ``result_bundles`` list will be used.
+    :type name_metric: str
 
-        :Parameters: :class:`TestMetric` parameters
-        """
-        self.metrics[name] = TestMetric(data, units)
+    :param result: Optionally, force the ``self.result`` attribute to that
+        value. This is useful when the way of combining the result bundles is
+        not the default one, without having to make a whole new subclass.
+    :type result: Result
 
-    def display_and_exit(self) -> type(None):
-        print("Test result: {}".format(self))
-        if self:
-            sys.exit(0)
+    This is useful for some tests that are naturally decomposed in subtests.
+
+    .. note:: Metrics of aggregated bundles will always be shown, but can be
+        augmented with new metrics using the usual API.
+    """
+    def __init__(self, result_bundles, name_metric=None, result=None):
+        self.result_bundles = result_bundles
+        self.name_metric = name_metric
+        self.extra_metrics = {}
+        self._forced_result = result
+
+    @property
+    def result(self):
+        forced_result = self._forced_result
+        if forced_result is not None:
+            return forced_result
+
+        def predicate(combinator, result):
+            return combinator(
+                res_bundle.result is result
+                for res_bundle in self.result_bundles
+            )
+
+        if predicate(all, Result.UNDECIDED):
+            return Result.UNDECIDED
+        elif predicate(any, Result.FAILED):
+            return Result.FAILED
+        elif predicate(any, Result.PASSED):
```

Comment:
Shouldn't this be:
`elif predicate(all, Result.PASSED):`?

It makes more sense to me as:
```
if predicate(all, Result.PASSED):
    return Result.PASSED
 elif predicate(any, Result.FAILED):
    return Result.FAILED
 else:
    return Result.UNDECIDED
```

Reply:
The basic semantic of UNDECIDED is that it should be ignored. That means that a mix of PASSED and UNDECIDED should be interpreted as PASSED. The only time UNDECIDED "wins" is when there are only UNDECIDED, so there is no relevant data to be based on. 

In practice, we might want to be more cautious for some tests and treat a mix of passed and undecided as undecided. That said, the more results you aggregate, the higher the probability of having at least one decided. We don't want to end up with a test that returns "undecided" all the time.  This is one of the main argument against such aggregated tests, unless they are just heuristics on top of results that can be visualised in a different way (like test_freq_invariance and test_cpu_invariance).  If they can be visualised separately, we can keep an eye on the rate of undecided and make sure it stays within reasonable bounds.

We can introduce a parameter to the constructor to switch to one semantic or to another, but I'd rather avoid introducing too many tunables only supposed to enable writing tests that are to some degree broken by design.

Reply:
Okay, fair enough. For the moment my purpose was just to make sure that the logic in this PR is not broken. But it is intentional so it sounds good to me.